File: nodejs/test/e2e/session_fs_sqlite.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { DatabaseSync } from "node:sqlite";
import { MemoryProvider, VirtualProvider } from "@platformatic/vfs";
import { mkdtempSync, realpathSync } from "fs";
import { tmpdir } from "os";
import { join } from "path";
import { describe, expect, it } from "vitest";
import type { SessionFsReaddirWithTypesEntry } from "../../src/generated/rpc.js";
import {
    approveAll,
    CopilotSession,
    SessionEvent,
    type SessionFsConfig,
    type SessionFsProvider,
    type SessionFsFileInfo,
    type SessionFsSqliteQueryResult,
    type SessionFsSqliteQueryType,
} from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

const sessionStatePath =
    process.platform === "win32"
        ? "/session-state"
        : join(
              realpathSync(mkdtempSync(join(tmpdir(), "copilot-sqlite-state-"))),
              "session-state"
          ).replace(/\\/g, "/");

const sessionFsConfig: SessionFsConfig = {
    initialCwd: "/",
    sessionStatePath,
    conventions: "posix",
    capabilities: { sqlite: true },
};

describe("Session Fs SQLite", async () => {
    const provider = new MemoryProvider();
    /** Track which queries were received, per session */
    const sqliteCalls: { sessionId: string; queryType: string; query: string }[] = [];
    /** Per-session SQLite databases, keyed by session ID.
     *  Stored at describe scope so the database survives if the CLI
     *  re-creates the handler (e.g., on reconnect). */
    const sessionDbs = new Map<string, DatabaseSync>();

    const createSessionFsHandler = (session: CopilotSession) =>
        createTestSessionFsHandlerWithSqlite(session, provider, sqliteCalls, sessionDbs);

    // Helpers to build session-namespaced paths for direct provider assertions
    const p = (sessionId: string, path: string) =>
        `/${sessionId}${path.startsWith("/") ? path : "/" + path}`;

    const { copilotClient: client } = await createSdkTestContext({
        copilotClientOptions: { sessionFs: sessionFsConfig },
    });

    it(
        "should route SQL queries through the sessionFs sqlite handler",
        { timeout: 60000 },
        async () => {
            const session = await client.createSession({
                onPermissionRequest: approveAll,
                createSessionFsHandler,
            });

            // Ask the agent to create a table and insert data using the SQL tool
            await session.sendAndWait({
                prompt:
                    'Use the sql tool to create a table called "items" with columns id (TEXT PRIMARY KEY) and name (TEXT). ' +
                    'Then insert a row with id "a1" and name "Widget".',
            });

            // Verify the sqlite handler was called with the right operations
            const sessionCalls = sqliteCalls.filter((c) => c.sessionId === session.sessionId);
            expect(sessionCalls.length).toBeGreaterThan(0);
            expect(sessionCalls.some((c) => c.query.toUpperCase().includes("CREATE TABLE"))).toBe(
                true
            );
            expect(sessionCalls.some((c) => c.query.toUpperCase().includes("INSERT"))).toBe(true);

            // Verify queryType is set correctly
            expect(sessionCalls.some((c) => c.queryType === "exec")).toBe(true);
            expect(sessionCalls.some((c) => c.queryType === "run")).toBe(true);

            await session.disconnect();
        }
    );

    it(
        "should allow subagents to use SQL tool via inherited sessionFs",
        { timeout: 60000 },
        async () => {
            const session = await client.createSession({
                onPermissionRequest: approveAll,
                createSessionFsHandler,
            });

            const events: SessionEvent[] = [];
            session.on((event) => {
                events.push(event);
            });

            // Ask the agent to use the task tool to spawn a subagent that uses SQL
            await session.sendAndWait({
                prompt:
                    "Use the task tool to ask a task agent to do the following: " +
                    "Use the sql tool to run this query: INSERT INTO todos (id, title, status) VALUES ('subagent-test', 'Created by subagent', 'done')",
            });

            await session.disconnect();

            // Verify that the subagent's SQL queries were routed through the sessionFs sqlite handler
            const sessionCalls = sqliteCalls.filter((c) => c.sessionId === session.sessionId);
            const insertCalls = sessionCalls.filter((c) =>
                c.query.toUpperCase().includes("INSERT")
            );
            expect(insertCalls.length).toBeGreaterThan(0);

            // Verify that the sql tool execution in events.jsonl came from the subagent (has agentId)
            const buf = await provider.readFile(
                p(session.sessionId, `${sessionStatePath}/events.jsonl`)
            );
            const content = buf.toString("utf8");
            const lines = content.split("\n").filter(Boolean);
            const parsed = lines.map((line) => JSON.parse(line));
            const sqlToolEvents = parsed.filter(
                (e: { type?: string; data?: { toolName?: string } }) =>
                    e.type === "tool.execution_start" && e.data?.toolName === "sql"
            );
            expect(sqlToolEvents.length).toBeGreaterThan(0);
            expect(sqlToolEvents.every((e: { agentId?: string }) => !!e.agentId)).toBe(true);
        }
    );
});

function createTestSessionFsHandlerWithSqlite(
    session: CopilotSession,
    provider: VirtualProvider,
    sqliteCalls: { sessionId: string; queryType: string; query: string }[],
    sessionDbs: Map<string, DatabaseSync>
): SessionFsProvider {
    const sp = (path: string) => `/${session.sessionId}${path.startsWith("/") ? path : "/" + path}`;

    function getOrCreateDb(): DatabaseSync {
        let db = sessionDbs.get(session.sessionId);
        if (!db) {
            db = new DatabaseSync(":memory:");
            db.exec("PRAGMA busy_timeout = 5000");
            sessionDbs.set(session.sessionId, db);
        }
        return db;
    }

    return {
        async readFile(path: string): Promise<string> {
            return (await provider.readFile(sp(path), "utf8")) as string;
        },
        async writeFile(path: string, content: string): Promise<void> {
            await provider.writeFile(sp(path), content);
        },
        async appendFile(path: string, content: string): Promise<void> {
            await provider.appendFile(sp(path), content);
        },
        async exists(path: string): Promise<boolean> {
            return provider.exists(sp(path));
        },
        async stat(path: string): Promise<SessionFsFileInfo> {
            const st = await provider.stat(sp(path));
            return {
                isFile: st.isFile(),
                isDirectory: st.isDirectory(),
                size: st.size,
                mtime: new Date(st.mtimeMs).toISOString(),
                birthtime: new Date(st.birthtimeMs).toISOString(),
            };
        },
        async mkdir(path: string, recursive: boolean, mode?: number): Promise<void> {
            await provider.mkdir(sp(path), { recursive, mode });
        },
        async readdir(path: string): Promise<string[]> {
            return (await provider.readdir(sp(path))) as string[];
        },
        async readdirWithTypes(path: string): Promise<SessionFsReaddirWithTypesEntry[]> {
            const names = (await provider.readdir(sp(path))) as string[];
            return Promise.all(
                names.map(async (name) => {
                    const st = await provider.stat(sp(`${path}/${name}`));
                    return {
                        name,
                        type: st.isDirectory() ? ("directory" as const) : ("file" as const),
                    };
                })
            );
        },
        async rm(path: string): Promise<void> {
            await provider.unlink(sp(path));
        },
        async rename(src: string, dest: string): Promise<void> {
            await provider.rename(sp(src), sp(dest));
        },
        sqlite: {
            async query(
                queryType: SessionFsSqliteQueryType,
                query: string,
                params?: Record<string, string | number | null>
            ): Promise<SessionFsSqliteQueryResult | undefined> {
                sqliteCalls.push({ sessionId: session.sessionId, queryType, query });

                const database = getOrCreateDb();
                const trimmed = query.trim();
                if (trimmed.length === 0) {
                    return undefined;
                }

                switch (queryType) {
                    case "exec":
                        database.exec(trimmed);
                        return undefined;

                    case "query": {
                        const stmt = database.prepare(trimmed);
                        const rows = (params ? stmt.all(params) : stmt.all()) as Record<
                            string,
                            unknown
                        >[];
                        const columns = rows.length > 0 ? Object.keys(rows[0]) : [];
                        return { rows, columns, rowsAffected: 0 };
                    }

                    case "run": {
                        const stmt = database.prepare(trimmed);
                        const result = params ? stmt.run(params) : stmt.run();
                        return {
                            rows: [],
                            columns: [],
                            rowsAffected: Number(result.changes),
                            lastInsertRowid:
                                result.lastInsertRowid !== undefined
                                    ? Number(result.lastInsertRowid)
                                    : undefined,
                        };
                    }
                }
            },
            async exists(): Promise<boolean> {
                return sessionDbs.has(session.sessionId);
            },
        },
    };
}

File: go/internal/e2e/session_fs_sqlite_e2e_test.go
package e2e

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"sync"
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
	"github.com/github/copilot-sdk/go/rpc"
)

type sqliteCall struct {
	SessionID string
	QueryType string
	Query     string
}

// inMemorySqliteProvider is a SessionFsProvider backed by in-memory maps with a stub SQLite handler.
// The stub returns plausible canned responses based on query type rather than executing real SQL.
// This avoids pulling in a real SQLite dependency (which would force a go directive bump across
// all scenario go.mod files).
type inMemorySqliteProvider struct {
	mu          sync.Mutex
	sessionID   string
	files       map[string]string
	dirs        map[string]bool
	hadQuery    bool
	sqliteCalls *[]sqliteCall
}

func newInMemorySqliteProvider(sessionID string, calls *[]sqliteCall) *inMemorySqliteProvider {
	return &inMemorySqliteProvider{
		sessionID:   sessionID,
		files:       make(map[string]string),
		dirs:        map[string]bool{"/": true},
		sqliteCalls: calls,
	}
}

func (p *inMemorySqliteProvider) ensureParent(path string) {
	parts := strings.Split(strings.TrimRight(path, "/"), "/")
	for i := 1; i < len(parts); i++ {
		p.dirs[strings.Join(parts[:i], "/")] = true
	}
}

func (p *inMemorySqliteProvider) ReadFile(path string) (string, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	content, ok := p.files[path]
	if !ok {
		return "", fmt.Errorf("file not found: %s", path)
	}
	return content, nil
}

func (p *inMemorySqliteProvider) WriteFile(path string, content string, mode *int) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.ensureParent(path)
	p.files[path] = content
	return nil
}

func (p *inMemorySqliteProvider) AppendFile(path string, content string, mode *int) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.ensureParent(path)
	p.files[path] = p.files[path] + content
	return nil
}

func (p *inMemorySqliteProvider) Exists(path string) (bool, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	_, isFile := p.files[path]
	_, isDir := p.dirs[path]
	return isFile || isDir, nil
}

func (p *inMemorySqliteProvider) Stat(path string) (*copilot.SessionFsFileInfo, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	now := time.Now().UTC()
	if p.dirs[path] {
		return &copilot.SessionFsFileInfo{
			IsFile: false, IsDirectory: true, Size: 0, Mtime: now, Birthtime: now,
		}, nil
	}
	if content, ok := p.files[path]; ok {
		return &copilot.SessionFsFileInfo{
			IsFile: true, IsDirectory: false, Size: int64(len(content)), Mtime: now, Birthtime: now,
		}, nil
	}
	return nil, fmt.Errorf("not found: %s", path)
}

func (p *inMemorySqliteProvider) MakeDirectory(path string, recursive bool, mode *int) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if recursive {
		parts := strings.Split(strings.TrimRight(path, "/"), "/")
		for i := 1; i <= len(parts); i++ {
			p.dirs[strings.Join(parts[:i], "/")] = true
		}
	} else {
		p.dirs[path] = true
	}
	return nil
}

func (p *inMemorySqliteProvider) ReadDirectory(path string) ([]string, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	prefix := strings.TrimRight(path, "/") + "/"
	names := map[string]bool{}
	for f := range p.files {
		if strings.HasPrefix(f, prefix) {
			rest := f[len(prefix):]
			if rest != "" {
				names[strings.SplitN(rest, "/", 2)[0]] = true
			}
		}
	}
	for d := range p.dirs {
		if strings.HasPrefix(d, prefix) {
			rest := d[len(prefix):]
			if rest != "" {
				names[strings.SplitN(rest, "/", 2)[0]] = true
			}
		}
	}
	result := make([]string, 0, len(names))
	for n := range names {
		result = append(result, n)
	}
	sort.Strings(result)
	return result, nil
}

func (p *inMemorySqliteProvider) ReadDirectoryWithTypes(path string) ([]rpc.SessionFsReaddirWithTypesEntry, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	prefix := strings.TrimRight(path, "/") + "/"
	entries := map[string]rpc.SessionFsReaddirWithTypesEntryType{}
	for d := range p.dirs {
		if strings.HasPrefix(d, prefix) {
			rest := d[len(prefix):]
			if rest != "" {
				name := strings.SplitN(rest, "/", 2)[0]
				entries[name] = rpc.SessionFsReaddirWithTypesEntryTypeDirectory
			}
		}
	}
	for f := range p.files {
		if strings.HasPrefix(f, prefix) {
			rest := f[len(prefix):]
			if rest != "" {
				name := strings.SplitN(rest, "/", 2)[0]
				if _, exists := entries[name]; !exists {
					entries[name] = rpc.SessionFsReaddirWithTypesEntryTypeFile
				}
			}
		}
	}
	result := make([]rpc.SessionFsReaddirWithTypesEntry, 0, len(entries))
	for name, typ := range entries {
		result = append(result, rpc.SessionFsReaddirWithTypesEntry{Name: name, Type: typ})
	}
	sort.Slice(result, func(i, j int) bool { return result[i].Name < result[j].Name })
	return result, nil
}

func (p *inMemorySqliteProvider) Remove(path string, recursive bool, force bool) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	delete(p.files, path)
	delete(p.dirs, path)
	return nil
}

func (p *inMemorySqliteProvider) Rename(src string, dest string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if content, ok := p.files[src]; ok {
		p.ensureParent(dest)
		p.files[dest] = content
		delete(p.files, src)
	}
	return nil
}

func (p *inMemorySqliteProvider) SqliteQuery(queryType rpc.SessionFsSqliteQueryType, query string, params map[string]any) (*copilot.SessionFsSqliteQueryResult, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.hadQuery = true
	*p.sqliteCalls = append(*p.sqliteCalls, sqliteCall{
		SessionID: p.sessionID,
		QueryType: string(queryType),
		Query:     query,
	})

	// Return canned results based on query type. The agent doesn't know or care
	// whether a real SQLite database is behind this — it just receives SQL tool
	// results. These stubs return plausible responses so the agent can proceed
	// normally without pulling in a real SQLite dependency.
	upper := strings.ToUpper(strings.TrimSpace(query))
	switch queryType {
	case rpc.SessionFsSqliteQueryTypeExec:
		return &copilot.SessionFsSqliteQueryResult{Columns: []string{}, Rows: []map[string]any{}}, nil
	case rpc.SessionFsSqliteQueryTypeRun:
		lastID := int64(1)
		return &copilot.SessionFsSqliteQueryResult{
			Columns:         []string{},
			Rows:            []map[string]any{},
			RowsAffected:    1,
			LastInsertRowid: &lastID,
		}, nil
	case rpc.SessionFsSqliteQueryTypeQuery:
		if strings.Contains(upper, "SELECT") {
			return &copilot.SessionFsSqliteQueryResult{
				Columns: []string{"id", "name"},
				Rows:    []map[string]any{{"id": "a1", "name": "Widget"}},
			}, nil
		}
		return &copilot.SessionFsSqliteQueryResult{Columns: []string{}, Rows: []map[string]any{}}, nil
	}
	return &copilot.SessionFsSqliteQueryResult{Columns: []string{}, Rows: []map[string]any{}}, nil
}

func (p *inMemorySqliteProvider) SqliteExists() (bool, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.hadQuery, nil
}

func TestSessionFsSqliteE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	sessionStatePath := createSessionStatePath(t)
	sessionFsConfig := &copilot.SessionFsConfig{
		InitialWorkingDirectory: "/",
		SessionStatePath:        sessionStatePath,
		Conventions:             rpc.SessionFsSetProviderConventionsPosix,
		Capabilities:            &copilot.SessionFsCapabilities{Sqlite: true},
	}

	var sqliteCalls []sqliteCall
	var providers sync.Map

	createSessionFsHandler := func(session *copilot.Session) copilot.SessionFsProvider {
		p := newInMemorySqliteProvider(session.SessionID, &sqliteCalls)
		providers.Store(session.SessionID, p)
		return p
	}

	client := ctx.NewClient(func(opts *copilot.ClientOptions) {
		opts.SessionFs = sessionFsConfig
	})
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should route sql queries through the sessionfs sqlite handler", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		sqliteCalls = nil

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest:     copilot.PermissionHandler.ApproveAll,
			CreateSessionFsProvider: createSessionFsHandler,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		msg, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: `Use the sql tool to create a table called "items" with columns id (TEXT PRIMARY KEY) and name (TEXT). ` +
				`Then insert a row with id "a1" and name "Widget".`,
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		_ = msg

		// Verify sqlite handler was called
		sessionCalls := filterCalls(sqliteCalls, session.SessionID)
		if len(sessionCalls) == 0 {
			t.Fatal("Expected sqlite handler to be called")
		}
		assertCallContains(t, sessionCalls, "CREATE TABLE")
		assertCallContains(t, sessionCalls, "INSERT")

		// Verify queryType is set correctly
		assertQueryType(t, sessionCalls, "exec")
		assertQueryType(t, sessionCalls, "run")

		if err := session.Disconnect(); err != nil {
			t.Fatalf("Failed to disconnect: %v", err)
		}
	})

	t.Run("should allow subagents to use sql tool via inherited sessionfs", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		sqliteCalls = nil

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest:     copilot.PermissionHandler.ApproveAll,
			CreateSessionFsProvider: createSessionFsHandler,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		_, err = session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Use the task tool to ask a task agent to do the following: " +
				"Use the sql tool to run this query: INSERT INTO todos " +
				"(id, title, status) VALUES ('subagent-test', 'Created by subagent', 'done')",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if err := session.Disconnect(); err != nil {
			t.Fatalf("Failed to disconnect: %v", err)
		}

		// Verify INSERT calls were routed
		sessionCalls := filterCalls(sqliteCalls, session.SessionID)
		insertCalls := filterByQuery(sessionCalls, "INSERT")
		if len(insertCalls) == 0 {
			t.Fatal("Expected INSERT calls from subagent")
		}

		// Read events.jsonl from in-memory FS
		val, ok := providers.Load(session.SessionID)
		if !ok {
			t.Fatal("Provider not found for session")
		}
		provider := val.(*inMemorySqliteProvider)
		eventsPath := sessionStatePath + "/events.jsonl"
		content, err := provider.ReadFile(eventsPath)
		if err != nil {
			t.Fatalf("Failed to read events.jsonl: %v", err)
		}
		lines := strings.Split(strings.TrimSpace(content), "\n")
		var sqlToolEvents []map[string]any
		for _, line := range lines {
			if line == "" {
				continue
			}
			var event map[string]any
			if err := json.Unmarshal([]byte(line), &event); err != nil {
				continue
			}
			if event["type"] == "tool.execution_start" {
				if data, ok := event["data"].(map[string]any); ok {
					if data["toolName"] == "sql" {
						sqlToolEvents = append(sqlToolEvents, event)
					}
				}
			}
		}
		if len(sqlToolEvents) == 0 {
			t.Fatal("Expected sql tool events in events.jsonl")
		}
		for _, e := range sqlToolEvents {
			if e["agentId"] == nil || e["agentId"] == "" {
				t.Error("Expected agentId on sql tool event")
			}
		}
	})
}

func filterCalls(calls []sqliteCall, sessionID string) []sqliteCall {
	var result []sqliteCall
	for _, c := range calls {
		if c.SessionID == sessionID {
			result = append(result, c)
		}
	}
	return result
}

func filterByQuery(calls []sqliteCall, keyword string) []sqliteCall {
	var result []sqliteCall
	for _, c := range calls {
		if strings.Contains(strings.ToUpper(c.Query), keyword) {
			result = append(result, c)
		}
	}
	return result
}

func assertCallContains(t *testing.T, calls []sqliteCall, keyword string) {
	t.Helper()
	for _, c := range calls {
		if strings.Contains(strings.ToUpper(c.Query), keyword) {
			return
		}
	}
	t.Errorf("Expected a call with query containing %q", keyword)
}

func assertQueryType(t *testing.T, calls []sqliteCall, queryType string) {
	t.Helper()
	for _, c := range calls {
		if c.QueryType == queryType {
			return
		}
	}
	t.Errorf("Expected a call with queryType %q", queryType)
}
